### DISCUSSION TITLE
How to override a manager for a specific file?

### DISCUSSION THREAD

**Discussion by chaayac** [renovatebot/renovate#25436] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

renovate/renovate:full

### If you're self-hosting Renovate, select which platform you are using.

None (running the Docker image)

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I'm trying to overwrite the manager used for specific dependencies in `go.mod` (hosted under a custom domain that points to a self-hosted Gitlab), so that specific packages will use the `gitlab-releases` datasource (with templating via the `regex` manager), while the others will use the default `gomod` manager (with `go` datasource).

### What you tried so far.

1. Disabling all the managers, only leaving `regex` enabled, but then it doesn't match against any files.
2. Using `customManagers` at the root of the config, still doesn't match anything.
2. Using `packageRules` with `customManagers` inside.

<details><summary>Latest (not working) config</summary>

```json
{
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "endpoint": "https://my.self-hosted.gitlab.domain/api/v4",
    "token": "<gitlab-token>",
    "platform": "gitlab",
    "gitUrl": "endpoint",
    "enabledManagers": [
        "regex"
    ],
    "customManagers": [
        {
            "customType": "regex",
            "matchStrings": [
                ".*my.go.proxy.domain/(?<path>.*?) (?<currentValue>.*?) .*"
            ],
            "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}",
            "datasourceTemplate": "gitlab-releases",
            "depNameTemplate": "{{{path}}}",
            "registryUrlTemplate": "https://my.self-hosted.gitlab.domain"
        }
    ],
    "packageRules": [
        {
            "matchFileNames": [
                "go.mod"
            ],
            "customManagers": [
                {
                    "customType": "regex",
                    "matchStrings": [
                        ".*my.go.proxy.domain/(?<path>.*?) (?<currentValue>.*?) .*"
                    ],
                    "versioningTemplate": "{{#if versioning}}{{{versioning}}}{{else}}semver{{/if}}",
                    "datasourceTemplate": "gitlab-releases",
                    "depNameTemplate": "{{{path}}}",
                    "registryUrlTemplate": "https://my.self-hosted.gitlab.domain"
                }
            ]
        }
    ],
    "repositories": [
        "my/custom/repo/patht"
    ]
}

```

</details> 

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

**Comment by rarkins** [comment]
This is not possible because the custom manager does not know how to update a go.sum afterwards. Why are you wanting to do this in the first place?

**Reply by chaayac** [reply]
The reason is that we don't actually use Gitlab for hosting the go packages, instead we have an in-house proxy to the Gitlab repository itself.

Anyway, I was able to get it working with this.

```json
"customManagers": [
        {
            "customType": "regex",
            "fileMatch": [
                "(^|/)go\\.mod$"
            ],
            "matchStrings": [
                "my\\.custom\\.domain\/(?<depName>.*?)(?<depv>\\/v\\d+)? (?<currentValue>[^\\s\\n]*)"
            ],
            "versioningTemplate": "semver",
            "datasourceTemplate": "gitlab-releases",
            "depNameTemplate": "{{{depName}}}",
            "registryUrlTemplate": "https://self.hosted.gitlab.domain"
        },
```

Yes, it doesn't update go.sum when there's only updates to our own internal packages. However there are (almost) always updates to external packages too, so luckily renovate runs go mod tidy and our packages go.sum entries get updated too :joy:.

Anyway, thanks for the attention and sorry for the delayed response.

**Reply by chaayac** [reply]
PS: I just realised I could probably combine it with `"postUpdateOptions": ["gomodTidy"]` for completeness too.
